Cargo.toml

```diff
@@ -15,6 +15,9 @@ build = "build.rs"
 documentation = "https://docs.esp-rs.org/esp-idf-svc/"
 rust-version = "1.77"
 
+[patch.crates-io]
+esp-idf-hal = { git = "https://github.com/esp-rs/esp-idf-hal", branch = "fix-timer-gpio-v4" }
```

Comment:
i merged the hal PR so that can be removed

Reply:
It has to stay. Only the "branch" part can be removed now. :( With a workspace setup, that would've been unnecessary.

Reply:
That's true.

Besides the point but just wanna say, since we had another issue with the added reset reasons that is now [fixed](https://github.com/esp-rs/esp-idf-hal/pull/444/files) in esp-idf-hal we might want to make a quick patch release for esp-idf-hal

Reply:
True. I've updated the changelog with both of these. Might release a new patch version shortly even though extending `ResetReason` with additional values is a breaking change. However, I guess we can consider the previous state as broken anyway? In any case we need `#[non_exhaustive]` on most of our `enum`s

Reply:
... which is a breaking change which would allow us to have less breaking changes in future. BTW non-exhaustive we need not just on enums, but also on `structs`. Likely on ALL of these.

Reply:
We would need "non_exhaustive" so we can "marked" it as a non-breaking change, though at the end we don't break things here. It just that esp-idf at the .patch level is adding stuff to C enums that is semi breaking stuff for us here, and a user faces it when he is actively is updating to the latest versions. Also non_exhausitve is not needed since we transpile the enum's from a list of const vars. The problem is that we are calling _ => panic in unknown variants ( which is i think overall the right call)  that is "breaking" stuff.

Reply:
> We would need "non_exhaustive" so we can "market" it as a non-breaking change, though at the end we don't break things here. It just that esp-idf at the .patch level is adding stuff to C enums that is semi breaking stuff for us here, and a user faces it when he is actively is updating to the latest ESP-IDF versions. 

ESP-IDF explicitly documents adding new members to a C struct / enum as a "non-breaking" change even if formally speaking it IS a breaking change. They even require a special style of initializing such a struct and switching on such an enum so that the user code does not break. Fortunately, on the Rust side, with `non_exhaustive` - for the **Rust** counterparts of these raw esp-idf-sys types we can enforce it in a clearer way.

> Also non_exhausitve is not needed since we transpile the enum's from a list of const vars. The problem is that we are calling _ => panic in unknown variants ( which is i think overall the right call) that is "breaking" stuff.

How **we** match on the raw `esp_idf_sys` constants / C structs when converting them to actual Rust enums / structs has nothing to do with `non_exhaustive`. `non_exhaustive` addresses how **the user** is matching on our own Rust level structs / enum. Specifically, when destructuring structs he should always use the `..` syntax, and when matching on enums he should always have a "catch all" `_ =>` variant, which then allows **us** to keep adding new values / members to the Rust struct / enum without this being a breaking change for the user anymore.